core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -1029,8 +1030,8 @@ default Optional<LimitApplicationResult<ConnectorTableHandle>> applyLimit(Connec
      */
     default Optional<ConstraintApplicationResult<ConnectorTableHandle>> applyFilter(ConnectorSession session, ConnectorTableHandle handle, Constraint constraint)
     {
-        if (constraint.getSummary().getDomains().isEmpty()) {
-            throw new IllegalArgumentException("constraint summary is NONE");
+        if (constraint.getSummary().isNone() || FALSE.equals(constraint.getExpression())) {
+            throw new IllegalArgumentException("constraint is FALSE");
```

Comment:
The check here was for added for documentation purposes https://github.com/trinodb/trino/commit/83e144a825a2ee604cdba1ee34baffcbbd37388c
> Fortunately, the `applyFilter` is never invoked with NONE domain, and
should not be, since the engine knows what to do in such case without
asking the connector.

The `|| FALSE.equals(constraint.getExpression()` doesn't seem to be just for documentation purposes though, since there is no equivalent logic in `PushPredicateIntoTableScan`. 

If current PushPredicateIntoTableScan code guarantees that `FALSE.equals(constraint.getExpression()` will never be true, this can be as is, but requires a code comment. If we go that way, I would split the `if` into two independent ones (the old condition and the new condition do not deserve same type of explanation)

Reply:
> with that commit being first, do all the test pass?

The issue was with `NULL` pushdown, not with `FALSE`, all tests should pass regardless its position among the other commits. Sorry if it caused confusion, do you want me to open a separate PR for this change?
 I saw that we validated `summary` is non-`NONE` and thought this validation is not whole without checking that the other part of the constraint (the expression) is also non-`FALSE`.

> If current PushPredicateIntoTableScan code guarantees that FALSE.equals(constraint.getExpression() will never be true

There's no explicit check for that, but `DomainTranslator`'s implementation translates `FALSE` expressions into `TupleDomain.none()` (via `visitBooleanLiteral`) so the remaining expression shouldn't be `FALSE` and therefore the translated connector expression shouldn't be `FALSE` as well. Do you think I better add an explicit check within `PushPredicateIntoTableScan`?

> The check here was for added for documentation purposes https://github.com/trinodb/trino/commit/83e144a825a2ee604cdba1ee34baffcbbd37388c

> > Fortunately, the applyFilter is never invoked with NONE domain, and

I was thinking that if we don't want it to be invoked with `NONE` domain (which is reasonable - why would we want to pushdown `NONE` filter into the connector?), then we probably don't want to pushdown `FALSE` expressions as well. I don't see a separation here between documentation and validation. If we say for documentation purposes that `summary` shouldn't be `NONE` then why not say that expression shouldn't be `FALSE`? I guess we don't want inconsistent behavior between the two. As mentioned earlier, I can add an explicit check to `PushPredicateIntoTableScan` if you think it's necessary.

> We now have two triggering conditions for the exception.
> If the exception is raised, it would be good to know which one triggered.

Agree, I separated it into 2 different `if` clauses with different exception messages.